manifest/provider/import.go

```diff
@@ -0,0 +1,211 @@
+			Detail:   err.Error(),
+		})
+		return resp, nil
+	}
+	s.logger.Trace("[ImportResourceState]", "[API Resource]", ro)
+
+	objectType, err := s.TFTypeFromOpenAPI(ctx, gvk, false)
+	if err != nil {
+		resp.Diagnostics = append(resp.Diagnostics, &tfprotov5.Diagnostic{
+			Severity: tfprotov5.DiagnosticSeverityError,
+			Summary:  fmt.Sprintf("Failed to determine resource type from GVK: %s", gvk),
+			Detail:   err.Error(),
+		})
+		return resp, nil
+	}
+
+	fo := RemoveServerSideFields(ro.UnstructuredContent())
+	nobj, err := payload.ToTFValue(fo, objectType, tftypes.NewAttributePath())
+	if err != nil {
+		resp.Diagnostics = append(resp.Diagnostics, &tfprotov5.Diagnostic{
+			Severity: tfprotov5.DiagnosticSeverityError,
+			Summary:  "Failed to convert unstructured to tftypes.Value",
+			Detail:   err.Error(),
+		})
+		return resp, nil
+	}
+	nobj, err = morph.DeepUnknown(objectType, nobj, tftypes.NewAttributePath())
+	if err != nil {
+		resp.Diagnostics = append(resp.Diagnostics, &tfprotov5.Diagnostic{
+			Severity: tfprotov5.DiagnosticSeverityError,
+			Summary:  "Failed to backfill unknown values during import",
+			Detail:   err.Error(),
+		})
+		return resp, nil
+	}
+	s.logger.Trace("[ImportResourceState]", "[tftypes.Value]", nobj)
+
+	newState := make(map[string]tftypes.Value)
+	wftype := rt.(tftypes.Object).AttributeTypes["wait_for"]
+	newState["manifest"] = tftypes.NewValue(tftypes.Object{AttributeTypes: map[string]tftypes.Type{}}, nil)
+	newState["object"] = morph.UnknownToNull(nobj)
+	newState["wait_for"] = tftypes.NewValue(wftype, nil)
+	nsVal := tftypes.NewValue(rt, newState)
+
+	impState, err := tfprotov5.NewDynamicValue(nsVal.Type(), nsVal)
+	if err != nil {
+		resp.Diagnostics = append(resp.Diagnostics, &tfprotov5.Diagnostic{
+			Severity: tfprotov5.DiagnosticSeverityError,
+			Summary:  "Failed to construct dynamic value for imported state",
+			Detail:   err.Error(),
+		})
+		return resp, nil
+	}
+	resp.ImportedResources = append(resp.ImportedResources, &tfprotov5.ImportedResource{
+		TypeName: req.TypeName,
+		State:    &impState,
+	})
+	resp.Diagnostics = append(resp.Diagnostics, &tfprotov5.Diagnostic{
+		Severity: tfprotov5.DiagnosticSeverityWarning,
+		Summary:  "Apply needed after 'import'",
+		Detail:   "Please run apply after a successful import to realign the resource state to the configuration in Terraform.",
+	})
+	return resp, nil
+}
+
+// parseImportID processes the resource ID string passed by the user to the "terraform import" command
+// and extracts the values for GVK, name and (optionally) namespace of the target resource as required
+// during the import process.
+//
+// The expected format for the import resource ID is:
+//
+// "apiVersion=<value>,kind=<value>,name=<value>[,namespace=<value>"]
+//
+// where 'namespace' is only required for resources that expect a namespace.
+//
+// Example: "apiVersion=v1,kind=Secret,namespace=default,name=default-token-qgm6s"
```

Comment:
It occurred to me that if the ordering of the keys is fixed we could just write a regex with named groups instead of having to parse the tokens and figure out if it's valid or not. Like this:

```go
func parseImportID(id string) (schema.GroupVersionKind, string, string, error) {
	re := regexp.MustCompile(
		`apiVersion=(?P<apiVersion>[-\w]+),` +
			`kind=(?P<kind>[-\w]+),` +
			`name=(?P<name>[-\w]+)` +
			`(,namespace=(?P<namespace>[-\w]+))?`)
	matches := re.FindStringSubmatch(id)
	if len(matches) == 0 {
		return schema.GroupVersionKind{}, "", "", fmt.Errorf("invalid import ID: expected format: apiVersion=<value>,kind=<value>,name=<value>[,namespace=<value>]")
	}

	m := map[string]string{"namespace": "default"}
	for _, k := range []string{"apiVersion", "kind", "name", "namespace"} {
		if v := matches[re.SubexpIndex(k)]; v != "" {
			m[k] = v
		}
	}

	gvk := schema.FromAPIVersionAndKind(m["apiVersion"], m["kind"])
	name := m["name"]
	namespace := m["namespace"]
	return gvk, name, namespace, nil
}
```

Reply:
Just playing regex golf now, but something like this works if we want to allow any ordering of the key value pairs:

```go
func parseImportID(id string) (schema.GroupVersionKind, string, string, error) {
	re := regexp.MustCompile(
		`((apiVersion=(?P<apiVersion>[-\w]+))|` +
			`(kind=(?P<kind>[-\w]+))|` +
			`(name=(?P<name>[-\w]+))|` +
			`(namespace=(?P<namespace>[-\w]+)))`)
	matches := re.FindAllStringSubmatch(id, -1)
	if len(matches) < 3 {
		return schema.GroupVersionKind{}, "", "", fmt.Errorf("invalid import ID: expected format: apiVersion=<value>,kind=<value>,name=<value>[,namespace=<value>]")
	}

	m := map[string]string{"namespace": "default"}
	for _, mm := range matches {
		for _, k := range []string{"apiVersion", "kind", "name", "namespace"} {
			if v := mm[re.SubexpIndex(k)]; v != "" {
				m[k] = v
			}
		}
	}

	gvk := schema.FromAPIVersionAndKind(m["apiVersion"], m["kind"])
	name := m["name"]
	namespace := m["namespace"]
	return gvk, name, namespace, nil
}
```

Reply:
I can switch to the regexp version, I have no preference TBH.
But I'm pretty sure the multiple branch regexp in the second example is a bit slower than the simple iteration when tokenizing. Then again, we're not building real-time trading systems here, so I have no preference one way or the other.

Would you say the regexp version improves the code readability? I can't really tell for sure. I'm a fan of regexp myself, but I know a lot of folks aren't.

Reply:
Nah – it's probably easier for someone new to the code to reason about a simple parser it they aren't a ninja with regular expressions and have awareness of named and capture groups and so on. We can keep as is, I'm just a fan of regex for little things like this.

Reply:
TBH, regexp used to be my first choice for things like these too. Until I noticed a lot of push-back from the less-than-ninja collaborators and I was surprised by how prevalent that is.